wire-ios-sync-engine/Source/Use cases/ImportBackupUseCase/ImportBackupUseCase.swift

```diff
@@ -0,0 +1,192 @@
+//
+// Wire
+// Copyright (C) 2025 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+import WireCrypto
+import WireDataModel
+import WireLogging
+import WireSystem
+import ZipArchive
+
+struct ImportBackupUseCase: ImportBackupUseCaseProtocol {
+
+    let userSession: () -> UserSession?
+    let dispatchGroup: ZMSDispatchGroup
+    let streamDecryptor: ImportBackupStreamDecryptorProtocol
+    let fileArchiver: ImportBackupFileArchiverProtocol
+    let entityStorage: ImportBackupEntityStorageProtocol
+    let appStateUpdater: ImportBackupAppStateUpdaterProtocol
+
+    let sharedContainerURL: URL
+    let logger: WireLogger
+
+    private let workerQueue = DispatchQueue(label: "import-backup")
+
+    func invoke(url: URL, password: String) async throws {
+
+        switch BackupFileExtensions(rawValue: url.pathExtension.lowercased()) {
+
+        case .fileExtensionWithUnderscore, .fileExtensionWithHyphen:
+            try await importIOSBackup(url, password)
+
+        case nil:
+            throw BackupRestoreError.invalidFileExtension
+        }
+    }
+
+    private func importIOSBackup(_ url: URL, _ password: String) async throws {
+
+        // to start with we need an active user session, later the session will be torn down
+        weak var userSession = userSession()
+        guard let account = userSession?.contextProvider.account else {
+            throw BackupRestoreError.noActiveAccount
+        }
+
+        // before we start the first operation let the user know, the progress has started
+        appStateUpdater.reportImportProgress(progress: 0.5)
```

Comment:
I notice we only report progress once, should we also report progress after decrypt and unzip, and after the store is replaced?

Reply:
I added another call to report progress: b0a53f035590d86f2ab15ffcd86cb3d2c4b70503
However, as soon as we tear down the user session we also tear down all the authenticated UI and we show the migration screen.
Since this will not be the case for CPB I prefer to not implement a progress label there.